### ISSUE TITLE
We need more unit tests!

### ISSUE THREAD

**Issue by duncanmmacleod** [gwastro/gwin#11]
@vivienr, @cdcapano, @cmbiwer, the unit tests currently only hit [22% of the code in `gwin` package](https://coveralls.io/builds/16385133) (this deliberately does not include the scripts under `/bin/`). I would like to get that number up around 90% (ideally, >75% more realistically) before anyone embarks on a grand new development project as part of `gwin`.

I know how to write unit tests, but have no idea how to actually test the code that exists in gwin at the moment, so am reaching out for help in writing new tests. The base plan is as follows:

- each `class` (e.g. `GwinClass`) has its own test class (`TestGwinClass`) (think of `unittest.TestCase`)
- each method of each `class` has its own unit test method as part of the test class
- each function has its own test function
- each line of code in each function should be tested somewhere, we can simplify annoying loops and the like by using [`fixtures`](https://docs.pytest.org/en/latest/fixture.html) and [parametrisation](https://docs.pytest.org/en/latest/parametrize.html).

Basically, every object we define in the package should have a corresponding test object. To that end I will set up a PR that defines a test class for every class definition I can find, and hope that y'all can help me populate it with useful stuff.

Testing the scripts in `/bin/` is another matter altogether, and not something I want to get to right now.

**Comment by cdcapano** [comment]
@duncanmmacleod Some of the classes/functions have an `Examples` section in the documentation; I think these would offer some good unit tests. For example, in `GaussianLikelihood`:

https://github.com/gwastro/gwin/blob/master/gwin/likelihood.py#L756-L813

I'm not sure how to turn this into a unit test that gets run automatically though. Could you turn one of those examples above into a unit test? Then I can follow your lead in implementing new tests.

**Comment by duncanmmacleod** [comment]
@cdcapano, forgive my ignorance, I hadn't seen the examples. These will make a great start, so I'll go through and see if I can work up at least one of these into a proper unit test. Thanks.

**Comment by duncanmmacleod** [comment]
@cdcapano, my PR #14 now has an example for the `GaussianLikelihood` class adapted from the reference you sent earlier.

After looking at the coverage and the code a bit more thoroughly, some of the largest holes in terms of lines covered would be easily tested if we had examples of some of the files required to run a workflow, e.g.

- a configuration file
- an `InferenceFile` in txt and hdf5 formats

@cdcapano, @vivienr, @cmbiwer, can you provide some of those that we can include in the repo (with `git-lfs` if appropriate).

**Comment by duncanmmacleod** [comment]
Actually, for the inference files its cleaner to just write a file as part of the test, and validate it, then read the same file (so we don't have to store a written file in the repo).

However, that requires me understanding the format of the `samples` arrays. Can someone explain that to me (are they stored in a `astropy.table.Table`, or `numpy.ndarray`, or similar?)

If someone could post code of how to mock up a small table of samples, using random data, that would likely be enough.

**Comment by cdcapano** [comment]
@duncanmmacleod The `samples` arrays are just `numpy.ndarray`s. They can be multi dimensional, depending on if a sampler uses an ensemble of walkers (2D) and multiple temperatures (3D). The last dimension is always the number of iterations; so, e.g., if you had 10 temps, 5 walkers, and 30 iterations with two parameters `x` and `y`, then `samples` would have two datasets, `x` and `y`; each data set would be a `numpy.ndarray` with shape 10 x 5 x 30.

Anyway, this a long-winded way of saying: if you have an array `x` of random data, just do: `fp['samples/x'] = x`.
